src/v/cluster/metrics_reporter.h

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "cluster/fwd.h"
+#include "cluster/health_monitor_types.h"
+#include "cluster/members_table.h"
+#include "cluster/topic_table.h"
+#include "cluster/types.h"
+#include "http/client.h"
+#include "model/metadata.h"
+#include "model/timestamp.h"
+#include "raft/consensus.h"
+#include "utils/prefix_logger.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/sharded.hh>
+#include <seastar/core/sstring.hh>
+
+#include <cstdint>
+#include <vector>
+
+namespace cluster {
+
+class metrics_reporter {
+public:
+    struct node_disk_space {
+        uint64_t free;
+        uint64_t total;
+    };
+
+    struct node_metrics {
+        model::node_id id;
+        uint32_t cpu_count;
+        bool is_alive;
+        ss::sstring version;
+        std::vector<node_disk_space> disks;
+        uint64_t uptime_ms;
+    };
+
+    struct metrics_snapshot {
+        static constexpr int16_t version = 1;
+
+        ss::sstring cluster_uuid;
+        uint64_t cluster_creation_epoch;
+        uint32_t topic_count;
+        uint32_t partition_count;
+
+        std::vector<node_metrics> nodes;
+    };
+    static constexpr ss::shard_id shard = 0;
+
+    metrics_reporter(
+      consensus_ptr,
+      ss::sharded<members_table>&,
+      ss::sharded<topic_table>&,
+      ss::sharded<health_monitor_frontend>&,
+      ss::sharded<ss::abort_source>&);
+
+    ss::future<> start();
+    ss::future<> stop();
+
+private:
+    void report_metrics();
+    http::client::request_header make_header(const iobuf&);
+    ss::future<> do_report_metrics();
+    ss::future<result<metrics_snapshot>> build_metrics_snapshot();
+
+    ss::future<http::client> make_http_client();
+    ss::future<> try_initialize_cluster_info();
+
+    static constexpr std::string_view protocol = "https";
+    static constexpr std::string_view host = "m.rp.vectorized.io";
+    static constexpr std::string_view path = "v2";
```

Comment:
if we use a configuration value for this then it will be straight forward to write a duck tape test. a web server can be spun up in a few lines of python: https://docs.python.org/3/library/http.server.html and we could wrap that in a duck tape background service.

Reply:
Added test for metrics reporter in ducktape